doom.d/config.org
#+title: config
#+PROPERTY: header-args:emacs-lisp :lexical t

* Doom

#+begin_src  emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
;; (setq user-full-name "John Doe"
;;       user-mail-address "john@doe.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-symbol-font' -- for symbols
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:

(setq doom-font (font-spec :family "JetBrains Mono" :weight 'regular)
      doom-variable-pitch-font (font-spec :family "Fira Sans"))

;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; disable Doom's own theme setting mechanism
(setq doom-theme nil)

(load-theme 'doom-moonlight t t)
(load-theme 'doom-gruvbox-light t t)

;; toolkit-theme-set-functions will be called on startup
(add-hook 'toolkit-theme-set-functions
          (lambda (theme-name)
            (pcase theme-name
              ('dark (progn (enable-theme 'doom-moonlight) (disable-theme 'doom-gruvbox-light)))
              ('light (progn (enable-theme 'doom-gruvbox-light) (disable-theme 'doom-moonlight))))))

(setq display-line-numbers-type 'relative)

;; Do not access the clipboard when yanking by default
;; (unnecessary with evil mode as there are the */+ registers already)
(setq select-enable-clipboard nil)

;; Use the system trash whenever it makes sense
(setq delete-by-moving-to-trash t)

;; Use Ctrl-Shift-V, Ctrl-Shift-X, Ctrl-Shift-C for clipboard yank/kill/copy
(map! "C-S-v" #'clipboard-yank)
(map! "C-S-x" #'clipboard-kill-region)
(map! "C-S-c" #'clipboard-kill-ring-save)

;; Bind right mouse button to open the context menu (instead of the default
;; behaviour to expand the selection and then kill it)
(map! "<mouse-3>" #'menu-bar-open)

;; Capture to org-roam daily note with <Leader>-X instead of calling the default org-capture
;; TODO idk if the with-eval-after-load is necessary or harmless
(with-eval-after-load 'org-roam
  (map! :leader :desc "Org-roam Capture" :n "X" #'org-roam-dailies-capture-today))
#+end_src

* Treemacs

#+begin_src emacs-lisp
;; Expand using a single click in Treemacs instead of a double click
(with-eval-after-load 'treemacs
  (map! :map treemacs-mode-map [mouse-1] #'treemacs-single-click-expand-action))
#+end_src


* Magit

Magit's ~forge~ requires authentication.

#+begin_src emacs-lisp
;; Only use Secret Service API
(with-eval-after-load 'auth
  (setopt auth-sources '("secrets:session" "secrets:Login")))

;; https://gist.github.com/pelletier/dc2708d769f2225c7f3c033050e1ff18
;; Format expected by magit forge:
;; (secrets-create-item "Login" "Emacs Forge Github" "YOUR API TOKEN" :host "api.github.com" :user "USERNAME^forge")
#+end_src

* emacs-conflict

For resolving Syncthing conflicts in notes.

Called via ~M-x emacs-conflict-resolve-conflicts~.

#+begin_src emacs-lisp
(use-package emacs-conflict)

(with-eval-after-load 'emacs-conflict
  ;; Monkeypatch emacs-conflict to use the system trash when possible and use
  ;; kill-buffer more safely
  (advice-add 'emacs-conflict--resolve-conflict :override
   (lambda (conflict)
   (let* ((normal (emacs-conflict--get-normal-filename conflict)))
     (emacs-conflict--resolve-ediff
      (list conflict normal)
      (lambda ()
         (when (y-or-n-p "Delete conflict file and save the original file? ")
           (let* ((conflict-buffer (get-file-buffer conflict)))
             (when conflict-buffer
               (kill-buffer conflict-buffer))
             (delete-file conflict t)))))))))
#+end_src

* org-mode

#+begin_src emacs-lisp
;; no need to use with-eval-after-load here as it is a directory variable
(setq org-directory "~/Documents/Notes/org")

(defun org--get-has-deadline-scheduled-str ()
  (let* ((deadline (org-entry-get (point) "DEADLINE"))
         (scheduled (org-entry-get (point) "SCHEDULED"))
         (deadline-str (if deadline "D" " "))
         (scheduled-str (if scheduled "S" " ")))
    (concat deadline-str " " scheduled-str)))

(with-eval-after-load 'org
  ;; Automatic line breaks at 80 columns. org-mode seems to be smart enough to prevent it for code blocks (like this one) or headings, and only does it for ordinary text.
  (add-hook 'org-mode-hook (lambda () (setq-local fill-column 80)))
  (add-hook 'org-mode-hook 'auto-fill-mode)

  ;; Main thing is that the file name (%c) is no longer shown, as it is usually noise.
  (setq org-agenda-prefix-format
        '((agenda . "%?i%+5:e %?-12t% s")
          ;; cannot use %s or %t in todo list view, see:
          ;; https://stackoverflow.com/questions/58820073/s-in-org-agenda-prefix-format-doesnt-display-dates-in-the-todo-view
          ;; The full timestamp would take up too much space anyway. What I mostly
          ;; care about is whether the item has any deadline at all.
          (todo . "%?i%+5:e %-4(org--get-has-deadline-scheduled-str)")
          (tags . "%?i%+5:e %-4(org--get-has-deadline-scheduled-str)")
          (search . "%?i%+5:e %-4(org--get-has-deadline-scheduled-str)")))

  (setq org-todo-keywords
        '((sequence
           "TODO(t)" "PROJ(p)" "LOOP(r)" "STRT(s!/!)" "WAIT(w@/!)" "HOLD(h)" "IDEA(i)"
           "|" "DONE(d!)" "KILL(k@)")
          (sequence "MAYBE(m)" "|" "YES(y@)" "NO(n@)")))

  (setq org-todo-keyword-faces
        '(("STRT" . +org-todo-active)
          ("WAIT" . +org-todo-onhold)
          ("HOLD" . +org-todo-onhold)
          ("MAYBE" . +org-todo-onhold)
          ("PROJ" . +org-todo-project)
          ("NO" . +org-todo-cancel)
          ("KILL" . +org-todo-cancel)
          ("IDEA" . +org-todo-onhold))))
#+end_src

** IDEA [#C] Icons via org-agenda-category-icon-alist :emacs:

** roam-agenda

Lists org-roam TODOs in org-agenda efficiently using org-roam's database.

Heavily copied from this pair of articles, with some additions like wrapping the
org-roam save hook in a minor mode:

- [[https://magnus.therning.org/2021-03-14-keeping-todo-items-in-org-roam.html][Keeping todo items in org-roam]]
- [[https://magnus.therning.org/2021-07-23-keeping-todo-items-in-org-roam-v2.html][Keeping todo items in org-roam v2]]

*** org-agenda part

#+begin_src emacs-lisp
;; Every task will have this tag as it will be inherited from filetags, so it should be ignored
(setq org-agenda-hide-tags-regexp "^todo$")

(defun roam-agenda--todo-files ()
  "Return a list of roam files containing the todo filetag"
  (org-roam-db-sync)
  (let ((todo-nodes (seq-filter (lambda (n)
                                  (seq-contains-p (org-roam-node-tags n) "todo"))
                                (org-roam-node-list))))
    (seq-uniq (seq-map #'org-roam-node-file todo-nodes))))

(defun roam-agenda--update-todo-files (&rest _)
  "Set org-agenda-files to all files in org-roam-directory that contain a 'todo' filetag"
  (setq org-agenda-files (roam-agenda--todo-files))
  ;; also add some select non-roam files in org-directory
  (push "Inbox.org" org-agenda-files) ; orgzly
  ;; this file
  (push "~/.config/home-manager/doom.d/config.org" org-agenda-files))

;; Reconstruct org-agenda-files every time an agenda command is dispatched
(with-eval-after-load 'org
  (advice-add 'org-agenda :before #'roam-agenda--update-todo-files))
#+end_src

*** org-roam part

#+begin_src emacs-lisp
(defun roam-agenda--get-filetags ()
  (split-string-and-unquote (or (cdr (car (org-collect-keywords '("filetags") '("filetags")))) "")))

(defun roam-agenda--add-filetag (tag)
  (let* ((new-tags (cons tag (roam-agenda--get-filetags)))
         (new-tags-str (combine-and-quote-strings new-tags)))
    (org-roam-set-keyword "filetags" new-tags-str))
  )

(defun roam-agenda--del-filetag (tag)
  (let* ((new-tags (seq-difference (roam-agenda--get-filetags) `(,tag)))
         (new-tags-str (combine-and-quote-strings new-tags)))
    (org-roam-set-keyword "filetags" new-tags-str)))

(defun roam-agenda--todo-p ()
  "Return non-nil if the current buffer has a TODO entry"
  (org-element-map
      (org-element-parse-buffer 'headline)
      'headline
    (lambda (h)
      (eq (org-element-property :todo-type h)
          'todo))
    nil 'first-match))

(defun roam-agenda--update-todo-tag ()
  "Update todo tag in the current buffer"
  (when (and
         ;; unsure why checking (not (active-minibuffer-window)) is necessary,
         ;; including this anyway because it's in the article
         (not (active-minibuffer-window))
         (org-roam-file-p))
    ;; the original version from the article uses `org-with-point-at 1` here, but
    ;; from the same article: "I suspect the use of `org-with-point-at 1` is unnecessary"
    (let* ((tags (roam-agenda--get-filetags))
           (is-todo (roam-agenda--todo-p)))
      (cond ((and is-todo (not (seq-contains-p tags "todo")))
             (roam-agenda--add-filetag "todo"))
            ((and (not is-todo) (seq-contains-p tags "todo"))
             (roam-agenda--del-filetag "todo"))))))

(define-minor-mode roam-agenda-mode
  "Minor mode that automatically adds 'todo' filetags to org-roam files with TODO headings on save"
  :init-value nil
  (if roam-agenda-mode
      (add-hook 'before-save-hook #'roam-agenda--update-todo-tag nil t)
    (remove-hook 'before-save-hook #'roam-agenda--update-todo-tag t)))

(with-eval-after-load 'org
  ;; note that this will be toggled together with org-mode, i.e.
  ;; org-mode <=> roam-agenda-mode
  (add-hook 'org-mode-hook #'roam-agenda-mode))
#+end_src

* reverse-im

#+begin_src emacs-lisp
;; Russian keyboard layout compatibility
(use-package reverse-im
  :hook doom-after-init
  :custom
  (reverse-im-input-methods '("russian-computer"))
  ;; Fix commands using custom read-char input dispatchers
  (reverse-im-read-char-advice-function #'reverse-im-read-char-include)
  (reverse-im-cache-file (locate-user-emacs-file "reverse-im-cache.el")))
#+end_src

* Web

#+begin_src emacs-lisp
;; Why does this default to 2? Who knows!
(setopt typescript-ts-indent-offset 4)
#+end_src

* org-roam capture templates

At the end of the config because emacs' elisp parsing for matching parens and
completions dies somewhere here.

Doom Emacs adds some templates to org-capture-templates, but these are not
integrated with org-roam.

(the doom evil binding for capturing to today's daily note is ~SPC n r d T~)

#+begin_src emacs-lisp
(with-eval-after-load 'org
  ;; Default value of org-roam-dailies-capture-templates:
  ;; (("d" "default" entry "* %?" :target
  ;;   (file+head "%<%Y-%m-%d>.org" "#+title: %<%Y-%m-%d>\n")))
  (setq org-roam-dailies-capture-templates
        '(("t" "Personal todo"
           entry
           "* TODO %?\nSCHEDULED: %T"
           :target
           (file+head+olp "%<%Y-%m-%d>.org" "#+STARTUP: overview\n#+title: %<%Y-%m-%d>\n" ("TODOs")))
          ("d" "default"
           entry
           "* %?"
           :target
           (file+head "%<%Y-%m-%d>.org" "#+STARTUP: overview\n#+title: %<%Y-%m-%d>\n"))))
  ;; If you have more than one template, org-roam-dailies-goto-today will
  ;; ask you for the capture template by default. This overrides the
  ;; dailies-goto functions to never prompt for a template interactively
  ;; and to just assume the default (empty) template if one is missing.

  (advice-add #'org-roam-dailies-goto-today :filter-args
              (lambda (args) (apply (lambda (&optional keys) (list (or keys "d"))) args)))
  (advice-add #'org-roam-dailies-goto-tomorrow :filter-args
              (lambda (args) (apply (lambda (n &optional keys) (list n (or keys "d"))) args)))
  (advice-add #'org-roam-dailies-goto-yesterday :filter-args
              (lambda (args) (apply (lambda (n &optional keys) (list n (or keys "d"))) args)))
  (advice-add #'org-roam-dailies-goto-date :filter-args
              (lambda (args) (apply (lambda (&optional prefer-future keys) (list prefer-future (or keys "d"))) args)))

  ;; goto next/previous note don't do the thing and so do not need to be overridden

  ;; TODO change this, this is set to the default value at the moment
  (setq org-roam-capture-templates
        '(("d" "default" plain "%?" :target
          (file+head "%<%Y%m%d%H%M%S>-${slug}.org" "#+title: ${title}\n") :unnarrowed t))))
#+end_src


doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
;; (setq user-full-name "John Doe"
;;       user-mail-address "john@doe.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-symbol-font' -- for symbols
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:

(setq doom-font (font-spec :family "Fira Code" :weight 'semi-light)
      doom-variable-pitch-font (font-spec :family "Fira Sans"))

;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!


(setq doom-theme 'doom-one)

(setq display-line-numbers-type 'relative)

(setq org-directory "~/Documents/Notes/org")

;; Expand using a single click in Treemacs instead of a double click
(after! treemacs
  (map! :map treemacs-mode-map [mouse-1] #'treemacs-single-click-expand-action))

;; Russian keyboard layout compatibility
(use-package! reverse-im
  :hook doom-after-init
  :custom
  (reverse-im-input-methods '("russian-computer"))
  ;; Fix commands using custom read-char input dispatchers
  (reverse-im-read-char-advice-function #'reverse-im-read-char-include)
  (reverse-im-cache-file (locate-user-emacs-file "reverse-im-cache.el")))



;; List org-roam TODOs in org-agenda efficiently using org-roam's database
;;
;; Heavily copied from this pair of articles:
;; https://magnus.therning.org/2021-03-14-keeping-todo-items-in-org-roam.html
;; https://magnus.therning.org/2021-07-23-keeping-todo-items-in-org-roam-v2.html

(defun roam-agenda--todo-files ()
  "Return a list of roam files containing the todo filetag"
  (org-roam-db-sync)
  (let ((todo-nodes (seq-filter (lambda (n)
                                  (seq-contains-p (org-roam-node-tags n) "todo"))
                                (org-roam-node-list))))
    (seq-uniq (seq-map #'org-roam-node-file todo-nodes))))

(defun roam-agenda--update-todo-files (&rest _)
  "Set org-agenda-files to all files in org-roam-directory that contain a 'todo' filetag"
  (setq org-agenda-files (roam-agenda--todo-files)))

;; Reconstruct org-agenda-files every time an agenda command is dispatched
(advice-add 'org-agenda :before #'roam-agenda--update-todo-files)


(defun roam-agenda--get-filetags ()
  (split-string-and-unquote (or (cdr (car (org-collect-keywords '("filetags") '("filetags")))) "")))

(defun roam-agenda--add-filetag (tag)
  (let* ((new-tags (cons tag (roam-agenda--get-filetags)))
         (new-tags-str (combine-and-quote-strings new-tags)))
    (org-roam-set-keyword "filetags" new-tags-str))
  )

(defun roam-agenda--del-filetag (tag)
  (let* ((new-tags (seq-difference (roam-agenda--get-filetags) `(,tag)))
         (new-tags-str (combine-and-quote-strings new-tags)))
    (org-roam-set-keyword "filetags" new-tags-str)))

(defun roam-agenda--todo-p ()
  "Return non-nil if the current buffer has a TODO entry"
  (org-element-map
      (org-element-parse-buffer 'headline)
      'headline
    (lambda (h)
      (eq (org-element-property :todo-type h)
          'todo))
    nil 'first-match))

(defun roam-agenda--update-todo-tag ()
  "Update todo tag in the current buffer"
  (when (and
         ;; unsure why checking (not (active-minibuffer-window)) is necessary,
         ;; including this anyway because it's in the article
         (not (active-minibuffer-window))
         (org-roam-file-p))
    ;; the original version from the article uses `org-with-point-at 1` here, but
    ;; from the same article: "I suspect the use of `org-with-point-at 1` is unnecessary"
    (let* ((tags (roam-agenda--get-filetags))
           (is-todo (roam-agenda--todo-p)))
      (cond ((and is-todo (not (seq-contains-p tags "todo")))
             (roam-agenda--add-filetag "todo"))
            ((and (not is-todo) (seq-contains-p tags "todo"))
             (roam-agenda--del-filetag "todo"))))))


(define-minor-mode roam-agenda-mode
  "Minor mode that automatically adds 'todo' filetags to org-roam files with TODO headings on save"
  :init-value nil
  (if roam-agenda-mode
      (add-hook 'before-save-hook #'roam-agenda--update-todo-tag nil t)
    (remove-hook 'before-save-hook #'roam-agenda--update-todo-tag t)))

;; note that this will be toggled together with org-mode, i.e.
;; org-mode <=> roam-agenda-mode
(add-hook 'org-mode-hook #'roam-agenda-mode)


(setq org-agenda-hide-tags-regexp "^todo$")
